From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin O'Connor <koconnor@usfirst.org>
Date: Tue, 18 Nov 2014 11:04:25 -0500
Subject: [PATCH 0492/6262] Moves C++ global build properties to properties
 file to match Java. Adds host reachability check and static IP fallback Adds
 roboRIO Image check Adds JRE check for Java

Change-Id: I07f3a0863bde0ebec7e7d8f48270e45758bddba5
---
 .../resources/templates/build.properties      | 17 ------
 .../resources/templates/build.xml             |  1 +
 .../resources/cpp-zip/ant/build.properties    | 17 ++++++
 .../src/main/resources/cpp-zip/ant/build.xml  | 57 +++++++++++--------
 .../resources/java-zip/ant/build.properties   |  2 +
 .../src/main/resources/java-zip/ant/build.xml | 46 +++++++++++++--
 6 files changed, 95 insertions(+), 45 deletions(-)
 create mode 100644 eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.properties

diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.properties b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.properties
index f807678b4a4c58a5caf6144f5ff2079e1cef79dc..91004e36d0ba839931eacddb19adebd69b701dab 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.properties
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.properties
@@ -1,20 +1,3 @@
-# Deployment information
-username=admin
-password=
-deploy.dir=/home/lvuser
-deploy.kill.command=/usr/local/frc/bin/frcKillRobot.sh -t -r
-deploy.log.file=/var/local/natinst/log/FRC_UserProgram.log
-command.dir=/home/lvuser/
-
-# Libraries to use
-wpilib=${user.home}/wpilib/cpp/${cpp-version}
-wpilib.lib=${wpilib}/lib
-
-# Ant support
-wpilib.ant.dir=${wpilib}/ant
-jsch.jar=${wpilib.ant.dir}/jsch-0.1.50.jar
-classloadertask.jar=${wpilib.ant.dir}/ant-classloadertask.jar
-
 # Build information
 out=FRCUserProgram
 src.dir=src
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.xml b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.xml
index 22e948e5bacefdd05eb6f76150765a223aa3021b..82e794019489f6e27bf7f9e9cba40e14a835be7f 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.xml
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/build.xml
@@ -21,6 +21,7 @@
   
   <property file="${user.home}/wpilib/wpilib.properties"/>
   <property file="build.properties"/>
+  <property file="${user.home}/wpilib/cpp/${version}/ant/build.properties"/>
   
   <import file="${wpilib.ant.dir}/build.xml"/>
 
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.properties b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.properties
new file mode 100644
index 0000000000000000000000000000000000000000..c409af4cc317364f6c04560dc61faf79d710233f
--- /dev/null
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.properties
@@ -0,0 +1,17 @@
+# Deployment information
+username=admin
+password=
+deploy.dir=/home/lvuser
+deploy.kill.command=/usr/local/frc/bin/frcKillRobot.sh -t -r
+deploy.log.file=/var/local/natinst/log/FRC_UserProgram.log
+command.dir=/home/lvuser/
+
+# Libraries to use
+wpilib=${user.home}/wpilib/cpp/${cpp-version}
+wpilib.lib=${wpilib}/lib
+roboRIOAllowedImages=18
+
+# Ant support
+wpilib.ant.dir=${wpilib}/ant
+jsch.jar=${wpilib.ant.dir}/jsch-0.1.50.jar
+classloadertask.jar=${wpilib.ant.dir}/ant-classloadertask.jar
\ No newline at end of file
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.xml b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.xml
index 36384450d6659305df89ab9e7186aac164ea0d13..fc22f0d8bed5030f9fb34ff3af3061980eb1cae7 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.xml
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/src/main/resources/cpp-zip/ant/build.xml
@@ -20,10 +20,25 @@
 
   <target name="get-target-ip">
     <property name="target" value="roboRIO-${team-number}.local" />
-    <echo>Target: ${target}</echo>
+	<echo>Trying Target: ${target}</echo>
+	 <if>
+		<isreachable host="${target}"/>
+	  <then>
+		<echo>roboRIO found via mDNS</echo>
+	  </then>
+	  <else>
+		<math result="ip.upper" operand1="${team-number}" operation="/" operand2="100" datatype="int"/>
+		<math result="ip.lower" operand1="${team-number}" operation="%" operand2="100" datatype="int"/>
+		<property name="target" value="10.${ip-upper}.${ip.lower}.2"/>
+		<echo>roboRIO not found via mDNS, falling back to static address of ${target}</echo>
+		<assert name="roboRIOFound" message="roboRIO not found, please check that the roboRIO is connected, imaged and that the team number is set properly in Eclipse">
+			<isreachable host="${target}"/>
+		</assert>
+	  </else>
+	 </if>
   </target>
 
-  <target name="deploy" depends="get-target-ip" description="Deploy the progam and start it running.">
+  <target name="deploy" depends="get-target-ip, dependencies" description="Deploy the progam and start it running.">
     <sshexec host="${target}"
          username="${username}"
          password="${password}"
@@ -57,28 +72,6 @@
 		   command="/usr/local/frc/bin/frcKillRobot.sh"/>
   </target>
 
-  <target name="debug-deploy" depends="get-target-ip" description="Deploy the jar and start the program running in debug mode.">
-    <echo>[athena-deploy] Killing running program</echo>
-    <sshexec host="${target}"
-             username="${username}"
-             password="${password}"
-             trust="true"
-             command="killall FRCUserProgram; sleep 1"/>
-    <echo>[athena-debug-deploy] Copying code over.</echo>
-    <scp file="${dist.jar}" todir="${username}@${target}:${deploy.dir}"
-	 password="${password}" trust="true"/>
-    <scp file="${wpilib.ant.dir}/debugcppprogram" todir="${username}@${target}:${deploy.dir}"
-	 password="${password}" trust="true"/>
-    <scp file="${wpilib.ant.dir}/debugjavaprogram" todir="${username}@${target}:${deploy.dir}"
-	 password="${password}" trust="true"/>
-    <echo>[athena-debug-deploy] Starting program.</echo>
-    <sshexec host="${target}"
-	     username="${username}"
-	     password="${password}"
-	     trust="true"
-	     command="chmod a+x debug*program; ${deploy.debug.command}"/>
-  </target>
-
   <!-- Simulation support -->
   <target name="simulate">
     <parallel>
@@ -102,4 +95,20 @@
       </sequential>
     </parallel>
   </target>
+  
+  <target name="dependencies" depends="get-target-ip">
+    <property name="ant.enable.asserts" value="true"/>
+	<post to="http://${target}/nisysapi/server" logfile="sysProps.xml" verbose="false" encoding="UTF-16LE">
+		<prop name="Function" value="GetPropertiesOfItem"/>
+		<prop name="Plugins" value="nisyscfg"/>
+		<prop name="Items" value="system"/>
+	</post>
+	<loadfile srcFile="sysProps.xml" encoding="UTF-16LE" property="roboRIOSysValues"/>
+	<propertyregex property="roboRIOImage" input="${roboRIOSysValues}" regexp="FRC_roboRIO_2015_v([0-9]+)" select="\1" defaultValue="ImageRegExFail"/>
+	<assert message="roboRIO Image does not match plugin, allowed image version: ${roboRIOAllowedImages}">
+		<bool>
+			<contains string="${roboRIOAllowedImages}" substring="${roboRIOImage}"/>
+		</bool>
+	</assert>
+  </target>
 </project>
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
index 35ac1d0750d22cbfe08d25241122d831566f862b..b259e7fa85a435394219df24e2504627eca606fb 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
@@ -9,6 +9,7 @@ debug.flag.dir=/tmp/
 debug.flag.command=chown lvuser:ni ${debug.flag.dir}frcdebug
 command.dir=/home/lvuser/
 version=current
+roboRIOJRE.dir=/usr/local/frc/JRE
 
 # Libraries to use
 wpilib=${user.home}/wpilib/java/${version}
@@ -21,6 +22,7 @@ networktables.sources=${wpilib.lib}/NetworkTables-sources.jar
 #jnaerator.jar=${wpilib.lib}/jnaerator-runtime.jar
 #classpath=${wpilib.jar}:${networktables.jar}:${jna.jar}:${jnaerator.jar}
 classpath=${wpilib.jar}:${networktables.jar}
+roboRIOAllowedImages=18
 
 # Ant support
 wpilib.ant.dir=${wpilib}/ant
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
index 47b2fb0ee3bf651571bf6d36613d90f3715f9821..d457e689c628fbdbdf9131c8fcf8d78c2ef6ff14 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
@@ -20,7 +20,22 @@
 
   <target name="get-target-ip">
     <property name="target" value="roboRIO-${team-number}.local" />
-     <echo>Target: ${target}</echo>
+	<echo>Trying Target: ${target}</echo>
+	 <if>
+		<isreachable host="${target}"/>
+	  <then>
+		<echo>roboRIO found via mDNS</echo>
+	  </then>
+	  <else>
+		<math result="ip.upper" operand1="${team-number}" operation="/" operand2="100" datatype="int"/>
+		<math result="ip.lower" operand1="${team-number}" operation="%" operand2="100" datatype="int"/>
+		<property name="target" value="10.${ip-upper}.${ip.lower}.2"/>
+		<echo>roboRIO not found via mDNS, falling back to static address of ${target}</echo>
+		<assert name="roboRIOFound" message="roboRIO not found, please check that the roboRIO is connected, imaged and that the team number is set properly in Eclipse">
+			<isreachable host="${target}"/>
+		</assert>
+	  </else>
+	 </if>
   </target>
 
   <target name="compile" description="Compile the source code.">
@@ -69,8 +84,8 @@
     <!-- We're running a clean here to get around a known ant issue where it does not detected changed constant variables.
      To get around this, we're recompiling the entire project, which is not an issue for most teams. If this is an issue
      for you, you can remove the clean here, just be sure to do a full rebuild after you've changed any constants.
-     Reference: http://stackoverflow.com/questions/6430001/ant-doesnt-recompile-constants -->
-  <target name="deploy" depends="get-target-ip,clean,jar" description="Deploy the jar and start the program running.">
+     Reference: http://stackoverflow.com/questions/6430001/ant-doesnt-recompile-constants -->	
+  <target name="deploy" depends="clean,jar,get-target-ip,dependencies" description="Deploy the jar and start the program running.">
     <echo>[athena-deploy] Copying code over.</echo>
     <scp file="${dist.jar}" todir="${username}@${target}:${deploy.dir}" password="${password}" trust="true"/>
     <scp file="${wpilib.ant.dir}/robotCommand" todir="${username}@${target}:${command.dir}" password="${password}" trust="true"/>
@@ -89,7 +104,7 @@
        command="${deploy.log.command}"/>
   </target>
 
-  <target name="debug-deploy" depends="get-target-ip,jar" description="Deploy the jar and start the program running.">
+  <target name="debug-deploy" depends="jar,get-target-ip,dependencies" description="Deploy the jar and start the program running.">
     <echo>[athena-deploy] Copying code over.</echo>
     <scp file="${dist.jar}" todir="${username}@${target}:${deploy.dir}" password="${password}" trust="true"/>
   	<!-- The remoteDebugCommand file is used by /usr/local/frc/bin/frcRunRobot.sh on the roboRIO  -->
@@ -184,4 +199,27 @@
       </sequential>
     </parallel>
   </target>
+  
+  <target name="dependencies" depends="get-target-ip">
+    <property name="ant.enable.asserts" value="true"/>
+	<post to="http://${target}/nisysapi/server" logfile="sysProps.xml" verbose="false" encoding="UTF-16LE">
+		<prop name="Function" value="GetPropertiesOfItem"/>
+		<prop name="Plugins" value="nisyscfg"/>
+		<prop name="Items" value="system"/>
+	</post>
+	<loadfile srcFile="sysProps.xml" encoding="UTF-16LE" property="roboRIOSysValues"/>
+	<propertyregex property="roboRIOImage" input="${roboRIOSysValues}" regexp="FRC_roboRIO_2015_v([0-9]+)" select="\1" defaultValue="ImageRegExFail"/>
+	<assert message="roboRIO Image does not match plugin, allowed image version: ${roboRIOAllowedImages}">
+		<bool>
+			<contains string="${roboRIOAllowedImages}" substring="${roboRIOImage}"/>
+		</bool>
+	</assert>
+	<echo>Checking for JRE. If this fails install the JRE using these instructions: http://wpilib.screenstepslive.com/s/4485/m/13809/l/243933-installing-java-8-on-the-roborio-java-only</echo>
+	<sshexec host="${target}"
+        username="${username}"
+        password="${password}"
+        trust="true"
+		failonerror="true"
+        command="test -d ${roboRIOJRE.dir}"/>
+  </target>
 </project>
